h/static/scripts/controllers/form-controller.js

```diff
@@ -0,0 +1,204 @@
+'use strict';
+
+var Controller = require('../base/controller');
+var { findRefs, setElementState } = require('../util/dom');
+var submitForm = require('../util/submit-form');
+
+/**
+ * A controller which adds inline editing functionality to forms
+ */
+class FormController extends Controller {
+  constructor(element, options) {
+    super(element, options);
+
+    setElementState(this.refs.cancelBtn, {hidden: false});
+    this.refs.cancelBtn.addEventListener('click', event => {
+      event.preventDefault();
+      this.cancel();
+    });
+
+    // List of groups of controls that constitute each form field
+    this._fields = Array.from(element.querySelectorAll('.js-form-input'))
+      .map(el => {
+        var parts = findRefs(el);
+        return {container: el, input: parts.formInput};
+      });
+
+    this.on('focus', event => {
+      var field = this._fields.find(field => field.input === event.target);
```

Comment:
You really don't like maps, do you? 😉

Given this is the only place you use `this._fields`, and you're doing an O(N) lookup in it, it would make more sense to have `this._fields` be a mapping between the input element and its container `js-form-input` field.

Alternatively, and even more simply, you could just not maintain this data structure at all, and instead just find the parent `.js-form-input` field when you receive a focus event:

```javascript
var input = event.target;
var field = getClosestParent(event.target, '.js-form-input');

...

function getClosestParent(element, selector) {
  while ((element = element.parentNode) && element.nodeType == Node.ELEMENT_NODE) {
    if (element.matches(selector)) {
      return element;
    }
  }
  return null;
}
```